### PR TITLE
feat: Starting keploy docker from keploy binary

### PR DESCRIPTION
**Pull request by PranshuSrivastava**
## Related Issue
  - Previously an alias needed to be set to start keploy using docker, this caused some confusion as sometimes users would forget to set it and then not understand why Keploy was not starting.

#### Describe the changes you've made
This PR makes Keploy smart enough to detect if the user is providing a docker run command to keploy while using the keploy binary, in that case keploy starts the keploy docker container itself, and does not break.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

### CODE REVIEW

File: cmd/record.go [hunk 1]
```diff
@@ -208,6 +208,17 @@ func (r *Record) GetCmd() *cobra.Command {
...
 			}
+			//Check if app command starts with docker or sudo docker.
+			dockerRelatedCmd, dockerCmd := utils.IsDockerRelatedCmd(appCmd)
+			if !isDockerCmd && dockerRelatedCmd {
+				// r.logger.Error("Please run the command with sudo -E env PATH=$PATH keploy record -c \"your command\"")
+				isDockerCompose := false
+				if dockerCmd == "docker-compose" {
+					isDockerCompose = true
+				}
+				utils.UpdateKeployToDocker("record", appCmd, isDockerCompose, appContainer, buildDelay.String())
```
Comment (gouravkrosx) [inline cmd/record.go:219]: Here, you are not passing each flag information.

File: cmd/test.go [hunk 2]
```diff
@@ -211,6 +211,16 @@ func (t *Test) GetCmd() *cobra.Command {
...
 				return errors.New("missing required -c flag or appCmd in config file")
 			}
+			//Check if app command starts with docker or sudo docker.
+			dockerRelatedCmd, dockerCmd := utils.IsDockerRelatedCmd(appCmd)
+			if !isDockerCmd && dockerRelatedCmd {
+				isDockerCompose := false
+				if dockerCmd == "docker-compose" {
+					isDockerCompose = true
+				}
+				utils.UpdateKeployToDocker("test", appCmd, isDockerCompose, appContainer, buildDelay.String())
```
Comment (gouravkrosx) [inline cmd/test.go:221]: Here, also each flag information is not passed.

**Comment by gouravkrosx** [comment]
@PranshuSrivastava Please address the latest comments. And fix your **CLA**

**Review by gouravkrosx — CHANGES_REQUESTED** [review]

File: utils/utils.go [hunk 3]
```diff
@@ -104,3 +107,176 @@ func HandlePanic() {
...
+			return
+		}
+		dockerContext := strings.Split(strings.TrimSpace(string(out)), "\n")[0]
+		dockerContext = strings.Split(dockerContext, "\n")[0]
+		if dockerContext == "colima" {
+			logger.Info("Starting keploy in docker with colima context, as that is the current context.")
+		}else{
+			logger.Info("Starting keploy in docker with default context, as that is the current context.")
+		}
+		*keployAlias = "sudo docker run --pull always --name keploy-v2 -e BINARY_TO_DOCKER=true -p 16789:16789 --privileged --pid=host -it -v " + os.Getenv("PWD") + ":/files -v /sys/fs/cgroup:/sys/fs/cgroup -v debugfs:/sys/kernel/debug:rw -v /sys/fs/bpf:/sys/fs/bpf -v /var/run/docker.sock:/var/run/docker.sock -v " + os.Getenv("HOME") + "/.keploy-config:/root/.keploy-config -v " + os.Getenv("HOME") + "/.keploy:/root/.keploy --rm ghcr.io/keploy/keploy "
```
Comment (gouravkrosx) [inline utils/utils.go:196]: In the case of default docker context, the alias is correct but for colima, the alias is the same as for Linux.

**Review by gouravkrosx — CHANGES_REQUESTED** [review]

File: cmd/record.go [hunk 4]
```diff
@@ -234,28 +257,10 @@ func (r *Record) GetCmd() *cobra.Command {
 			} else {
 				// user provided the absolute path
 			}
-
```
[the remark this answers is not in the document]
Comment (PranshuSrivastava) [reply cmd/record.go:237]: done

File: cmd/record.go [hunk 5]
```diff
@@ -234,28 +257,10 @@ func (r *Record) GetCmd() *cobra.Command {
...
-			if isDockerCmd && buildDelay <= 30*time.Second {
-				r.logger.Warn(fmt.Sprintf("buildDelay is set to %v, incase your docker container takes more time to build use --buildDelay to set custom delay", buildDelay))
-				r.logger.Info(`Example usage: keploy record -c "docker-compose up --build" --buildDelay 35s`)
-			}
-
 			path += "/keploy"
 
 			r.logger.Info("", zap.Any("keploy test and mock path", path))
 
-			var hasContainerName bool
```
[the remark this answers is not in the document]
Comment (PranshuSrivastava) [reply cmd/record.go:247]: done

File: cmd/test.go [hunk 6]
```diff
@@ -308,6 +308,40 @@ func (t *Test) GetCmd() *cobra.Command {
 				}
 			}
 
+			mongoPassword, err := cmd.Flags().GetString("mongoPassword")
+			if err != nil {
+				t.logger.Error("failed to read the ports of outgoing calls to be ignored")
```
[the remark this answers is not in the document]
Comment (PranshuSrivastava) [reply cmd/test.go:313]: done

File: keploy-config.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
+record:
```
[the remark this answers is not in the document]
Comment (PranshuSrivastava) [reply keploy-config.yaml:1]: done

File: utils/utils.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PranshuSrivastava) [reply utils/utils.go:196]: done

File: utils/utils.go [hunk 8]
```diff
@@ -104,3 +107,176 @@ func HandlePanic() {
...
+			keployAlias = keployAlias + " --passThroughPorts " + fmt.Sprintf("%v ", joinedPorts)
+		}
+		if recordFlags.ConfigPath != "." {
+			keployAlias = keployAlias + " --configPath " + recordFlags.ConfigPath
+		}
+		if len(recordFlags.Path) > 0 {
+			keployAlias = keployAlias + " --path " + recordFlags.Path
+		}
+		addtionalFlags := " --containerName " + recordFlags.ContainerName + " --buildDelay " + recordFlags.BuildDelay.String() + " --delay " + fmt.Sprintf("%d", recordFlags.Delay) + " --proxyport " + fmt.Sprintf("%d", recordFlags.Proxyport) + " --networkName " + recordFlags.NetworkName + " --enableTele=" + fmt.Sprintf("%v", recordFlags.EnableTele)
+		if isDockerCompose {
```
[the remark this answers is not in the document]
Comment (PranshuSrivastava) [reply utils/utils.go:236]: removed

File: utils/utils.go [hunk 9]
```diff
@@ -104,3 +107,176 @@ func HandlePanic() {
...
+			keployAlias = keployAlias + " --configPath " + testFlags.ConfigPath
+		}
+		if len(testFlags.Testsets) > 0 {
+			keployAlias = keployAlias + " --testsets " + fmt.Sprintf("%v", testFlags.Testsets)
+		}
+		if len(testFlags.Path) > 0 {
+			keployAlias = keployAlias + " --path " + testFlags.Path
+		}
+		addtionalFlags := " --containerName " + testFlags.ContainerName + " --buildDelay " + testFlags.BuildDelay.String() + " --delay " + fmt.Sprintf("%d", testFlags.Delay) + " --networkName " + testFlags.NetworkName + " --enableTele=" + fmt.Sprintf("%v", testFlags.EnableTele) + " --apiTimeout " + fmt.Sprintf("%d", testFlags.ApiTimeout) + " --mongoPassword " + testFlags.MongoPassword + " --coverageReportPath " + testFlags.CoverageReportPath + " --withCoverage " + fmt.Sprintf("%v", testFlags.WithCoverage) + " --proxyport " + fmt.Sprintf("%d", testFlags.Proxyport)
+		if isDockerCompose {
```
[the remark this answers is not in the document]
Comment (PranshuSrivastava) [reply utils/utils.go:263]: removed

File: utils/utils.go (hunk 3 above)
Comment (PranshuSrivastava) [reply utils/utils.go]: done

**Pull request closed without merging**